rfc/data-privacy/v2.md

```diff
@@ -0,0 +1,637 @@
+For obfuscation policies to work, we need to define context objects that contain all the information necessary to evaluate the expressions:
+
+### Response Field Context Object:
+
+```json
+{
+  "fieldName": "email",
+  "fieldType": "String",
+  "parentType": "User",
+  "value": "user@example.com",
+  "valueDataType": "string"
+}
+```
+
+### Request Context Object:
+
+The "activate" expression has access to the request context object.
+The response context object is defined here: https://cosmo-docs.wundergraph.com/router/configuration/template-expressions
+
+## Configuration Structure
+
+The data privacy configuration consists of:
+
+1. **Transformers**: Reusable obfuscation logic defined using expr-lang
+2. **Policies**: Rules for when and how to apply obfuscation
+3. **Configuration Options**: Options to control behavior and validation
+
+```yaml
+data_privacy:
+  obfuscation:
+    # Configuration options
+    options:
+      undefined_field_behavior: "warn" # Options: "warn", "error", "ignore"
+      validate_transformers: true # Validate transformer return types on startup
+    
+    # Reusable transformers
+    transformers:
+      - name: "email"
+        expression: |
+          repeat("*", len(split(value, "@")[0])) + "@" + split(value, "@")[1]
+      - name: "default"
+        expression: |
+          valueDataType == "boolean" ? false : 
+          valueDataType == "number" ? 0 : 
+          repeat("*", len(value))
+      - name: "isoDate"
+        expression: |
+          date(value, "2006-01-02 15:04:05").Year() + "-**-**"
+    
+    # Obfuscation policies
+    policies:
+      - name: "Data Scientist Obfuscation"
+        # When this policy is active (evaluated once per request)
+        activate: "'data-scientist' in request.auth.roles"
+        
+        # YAML-based targeting (preferred for readability)
+        # Note: targets are evaluated first and take precedence over custom rules
+        targets:
+          - type: "User"
+            fields: ["email"]
+            transform: "email"
+          - type: "User" 
+            fields: ["phoneNumber"]
+            transform: "default"
+          - scalar_type: "ISODate"
+            transform: "isoDate"
+        
+        # For complex conditions not easily expressed in YAML
+        # Custom rule is only evaluated if no target rule matches
+        custom: |
+          # This expression is evaluated for each field
+          # It can access the field context and apply custom logic
+          typeName == "LogEntry" && fieldName == "ipAddress" ? 
+            replace(value, "\\d+$", "0") : 
+            value
+      
+      - name: "Developer Obfuscation"
+        activate: "'developer' in request.auth.roles"
+        # Note: Either use a default_transform OR custom, not both.
+        # When default_transform is specified, custom rules will be ignored.
+        default_transform: "default" # Applied to all fields not matched by targets
+        targets:
+          - type: "User"
```

Comment:
In my opinion, marking fields as potentially sensitive data should be done during schema design. I recognize the pitfall of types becoming out of sync, and it is unreasonable to expect developers to anticipate that this configuration needs to be updated and tested after a schema change. To me, this falls under the category of schema governance.

However, I like the concept of polices and transformers. They can still be defined through configuration, and their activation can also be determined at runtime. This combines the benefits of both approaches.

```graphql
type User {
  email: String @openfed__obfuscate(policy: "Developer Obfuscation")
}
```